### ISSUE TITLE
Terminal cursor displaced after minimize

### ISSUE THREAD

**Issue by RenzoGerritzen** [lapce/lapce#3085]
## Lapce Version

0.3.1

## System information

Windows 11

## Describe the bug

If the content viewer has a file or tab opened of which the contents are taller than the view height, then you minimize the window, the cursor of the terminal will be in te wrong location.

## Additional information

Before minimizing:
![Screenshot 2024-03-18 143635](https://github.com/lapce/lapce/assets/60329202/86ebc9e2-d7d1-4d91-8a88-c36c20154190)

After minimizing and opening:
![Screenshot 2024-03-18 143629](https://github.com/lapce/lapce/assets/60329202/e5e16162-d14b-48e0-bd15-0c5ab44d5df5)

After typing:
![Screenshot 2024-03-18 143639](https://github.com/lapce/lapce/assets/60329202/e3bd777a-5eaa-40f2-ae3d-85c465143ef6)

Based on the small flickering of the window upon opening it, i think it might have something to do with a re-drawing of the element. As it only seems to be happening when you press the minimize button and not when activating another window, it might have something to do with a state change when ghe minimizing happens.

Is it alright if i pick up the issue and see if i can fix it?

**Comment by MinusGix** [comment]
> Is it alright if i pick up the issue and see if i can fix it?

Feel free to try fixing it!
You'd want to look at `lapce-app/src/terminal/view.rs and probably the `wheel_scroll` function in `terminal/data.rs`, since it seems like some issue with scroll changing but possibly not getting sent to the terminal? Unsure.

**Comment by RenzoGerritzen** [comment]
Thank you! I'll get started on it in a few days and i will keep you updated! Great work on the editor by the way. It seems to run smoothly and i can only imagine the amount of time and effort to get to this point

**Comment by RenzoGerritzen** [comment]
I found out that when the application is minimized, the view buffer of the window only shows a small portion of the titlebar. This can be seen for a split second when you open the application.
![image](https://github.com/lapce/lapce/assets/60329202/f11d4485-f8d5-4ae3-8cd2-e59b06b43ceb)

This caused the terminals size to become 0 or less, causing it to move the cursor to the incorrect location. When the window is opened, the terminal is resized again and the cursor is moved again, causing it to render at the wrong location.

Is the scaling of the application intentional to save performance? As this is a pretty minor bug i want to make sure the fix will not break more important components.

**Comment by MinusGix** [comment]
I'm unsure as to the answer here. It might be something done in floem?

**Comment by RenzoGerritzen** [comment]
I am not sure myself to be honest. I'm looking into it, but if it is something caused by floem it might have to take the problem upstream.

**Comment by Long0x0** [comment]
On Windows, minimize means resize. We need a patch for the `WM_SIZE` event in the upstream winit crate.

See:
- [Why do minimized windows have an apparent size of 160×31?
](https://devblogs.microsoft.com/oldnewthing/20050210-00/?p=36483)
- [WM_SIZE message](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-size)
- https://github.com/rust-windowing/winit/issues/2015

**Comment by RenzoGerritzen** [comment]
Thank you kindly! I will see if i can get some kind of workaround

**Comment by RenzoGerritzen** [comment]
It seems the only way to really stop the cursor from displacing is to keep the width and height of the element when there is a drastic change in the size or to check for the exact size the window is when minimized, but this seems quite hacky an unreliable. I am not entirely sure how to proceed, but i think trying to solve the problem in lapce itself might not be a great idea. I'll keep the issue opened for a while to see if anyone else has some ideas, but i will close the issue soon

**Comment by MinusGix** [comment]
(The issue should remain open even if we can't fix it right now — thank you for trying at it! — because it would still be a bug)
I think ideally this is something fixed, or given a workaround, in floem. So we could have a change in our winit fork to support this.

**Comment by RenzoGerritzen** [comment]
Good point. I didn't know lapce used a custom winit fork, I'll give it a go

**Comment by RenzoGerritzen** [comment]
After some diffing i found out that while there is a resize event taking place when minimizing, it does not seem to be manually called. The winit crate seems to simply set a flag to the window and windows does the rest. Even if i could capture the minimizing event and bypass the size change somehow, the bug still happens when the window is made very small.

The only solutions i can think of are a minimum size or bypassing the size setting when checking the percentage of window scaling. Both solutions don't seem reliable. If anyone has any ideas i am glad to try them, but for now i am not sure how to proceed.

**Comment by Long0x0** [comment]
I created a PR for winit https://github.com/lapce/winit/pull/6 to fix the issue when minimizing. But it doesn't work when manually resizing to a small size.

I believe that's a bug in ConPTY, as window terminal, vscode integrated terminal, and alacritty have similar issues.

https://github.com/microsoft/terminal/issues?q=is%3Aissue+is%3Aopen+label%3AProduct-Conpty+resize
